obspy/io/gcf/core.py

```diff
@@ -33,68 +142,584 @@ def merge_gcf_stream(st):
+   ss.50 (2/4), or ss.75 (3/4). Use function :func:get_time_denominator to get the 
+   associated denominator.
+
+   The GCF format is only guaranteed to support 32-bit signed integer values. While data with values
+   out of range may be properly stored in the GCF format (if first and last data sample can be represented
+   as a 32-bit signed integer as well as all first difference values of the dat avector) the current 
+   implementaion only permitts input data to be representable as a 32-bit signed integer. If input waveforms
+   cannot be representable as 32-bit signed integers they will silently be clipped at -2,147,483,648 and 
+   2,147,483,647
+     
+   ..rubric:: Exceptions
+   
+   :class:TypeError will be raised upon unsupported input
+   :class:IOError will be raised upon failure to output file
+   
+   ..rubric:: Example
+   
+   >>> st.write('GCF-filename.gcf', format='GCF') #doctest: +SKIP
+   
+   """
+   # Make sure we have a Stream or Trace object
+   if not isinstance(stream,Stream):
+      if not isinstance(stream,Trace):
+         raise TypeError("write requires either a Stream or a Trace object")
+      else:
+         stream = Stream(traces=stream)
+         
+   # Check output path
+   if os.path.exists(filename) and not os.path.isfile(filename):
+      IOError("path %s exists but is not a file object" % (filename))
+   d,f = os.path.split(filename)
+   if d and not os.path.exists(d):
+      try:
+         os.makedirs(d)
+      except:
+         raise IOError("failed to create output directorie(s) %s" % (d))
+         
+   # check missalign parameter for supported range
+   if missalign < 0 or missalign > 0.5:
+      raise TypeError("argument missalign out of range (permitted range 0 - 0.5)")
+   
+   ret = 0
+   # Load shared library
+   try:
+      gcf_io = _load_cdll("gcf")
+   except Exception as e:
+      raise IOError(str(e))
+   else:
+      # declare function argument and return types
+      gcf_io.write_gcf.argtypes = [ctypes.c_char_p, ctypes.POINTER(_GcfFile)]
+      gcf_io.write_gcf.restype = ctypes.c_int
+
+      # prepare the GcfFile object
+      obj         = _GcfFile()
+      obj.n_seg   = len(stream)
+      obj.seg     = (_GcfSeg * obj.n_seg)()
+      obj.n_alloc = obj.n_seg
+      ctypes.cast(obj.seg, ctypes.POINTER(_GcfSeg))
+
+      for i,trace in enumerate(stream):
+         gcf_stats = trace.stats.gcf if hasattr(trace.stats,'gcf') else None
+         
+         ### Input checks, all of these are checked in the C-writer function
+         ###  but still duplicated here for clarity and ease of formatting
+         ###  proper exception messages
+
+         ### Check data type
+         if not isinstance(trace.data[0],np.int32):
+            data = []
+            for d in trace.data:
+                if d < -2147483648:
+                    d = -2147483648
+                elif d > 2147483647:
+                    d = 2147483647
```

Comment:
would be good to set magic numbers at the top as constants and reuse here, I think, similar to the C header file. also can be cleaned up a bit

something like.. (naming just a suggestion)
```
SAMPLE_MAX = 2147483647
SAMPLE_MIN = -2147483648

[...]

trace.data[trace.data < SAMPLE_MIN] = SAMPLE_MIN
...
```

or even better, use `np.ceil` and `np.floor`

Reply:
see https://github.com/obspy/obspy/pull/3055#discussion_r859630213

Reply:
Will do (one of the two)